executor/benchmark_test.go

```diff
@@ -0,0 +1,344 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sort"
+	"testing"
+
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/mock"
+)
+
+var (
+	_ Executor = &mockDataSource{}
+)
+
+type mockDataSourceParameters struct {
+	schema *expression.Schema
+	ndvs   []int  // number of distinct values on columns[i] and zero represents no limit
+	orders []bool // columns[i] should be ordered if orders[i] is true
+	rows   int    // number of rows the DataSource should output
+	ctx    sessionctx.Context
+}
+
+type mockDataSource struct {
+	baseExecutor
+	p        mockDataSourceParameters
+	colData  [][]interface{}
+	chunks   []*chunk.Chunk
+	chunkPtr int
+}
+
+func (mds *mockDataSource) genColDatums(ordinal int) (results []interface{}) {
+	typ := mds.retFieldTypes[ordinal]
+	order := mds.p.orders[ordinal]
+	rows := mds.p.rows
+	NDV := mds.p.ndvs[ordinal]
+	results = make([]interface{}, 0, rows)
+	if NDV == 0 {
+		for i := 0; i < rows; i++ {
+			results = append(results, mds.randDatum(typ))
+		}
+	} else {
+		datumSet := make(map[string]bool, NDV)
+		datums := make([]interface{}, 0, NDV)
+		for len(datums) < NDV {
+			d := mds.randDatum(typ)
+			str := fmt.Sprintf("%v", d)
+			if datumSet[str] {
+				continue
+			}
+			datumSet[str] = true
+			datums = append(datums, d)
+		}
+
+		for i := 0; i < rows; i++ {
+			results = append(results, datums[rand.Intn(NDV)])
+		}
+	}
+
+	if order {
+		sort.Slice(results, func(i, j int) bool {
+			switch typ.Tp {
+			case mysql.TypeLong, mysql.TypeLonglong:
+				return results[i].(int64) < results[j].(int64)
+			case mysql.TypeDouble:
+				return results[i].(float64) < results[j].(float64)
+			default:
+				panic("not implement")
+			}
+		})
+	}
+
+	return
+}
+
+func (mds *mockDataSource) randDatum(typ *types.FieldType) interface{} {
+	switch typ.Tp {
+	case mysql.TypeLong, mysql.TypeLonglong:
+		return int64(rand.Int())
+	case mysql.TypeDouble:
+		return rand.Float64()
+	default:
+		panic("not implement")
+	}
+}
+
+func (mds *mockDataSource) prepareChunks() {
+	mds.chunks = make([]*chunk.Chunk, (mds.p.rows+mds.initCap-1)/mds.initCap)
+	for i := range mds.chunks {
+		mds.chunks[i] = chunk.NewChunkWithCapacity(mds.retTypes(), mds.ctx.GetSessionVars().MaxChunkSize)
+	}
+
+	for i := 0; i < mds.p.rows; i++ {
+		idx := i / mds.initCap
+		types := mds.retTypes()
+		for colIdx := 0; colIdx < len(types); colIdx++ {
+			switch types[colIdx].Tp {
+			case mysql.TypeLong, mysql.TypeLonglong:
+				mds.chunks[idx].AppendInt64(colIdx, mds.colData[colIdx][i].(int64))
```

Comment:
we can use a `chunk.List` to store all the `mds.colData`, then we just need to do a deep copy for each column, it will be much faster.

Reply:
emm... I have considered about that too.. but there is no `Copy` method for `Chunk`.
I will add such a method.

Reply:
resolved